tests/bsim/bluetooth/audio/src/csip_notify_client_test.c

```diff
@@ -0,0 +1,198 @@
+};
+
+extern enum bst_result_t bst_result;
+
+CREATE_FLAG(flag_csip_set_lock_discovered);
+CREATE_FLAG(flag_all_notifications_received);
+
+static struct bt_uuid_16 uuid = BT_UUID_INIT_16(0);
+
+static struct csip_instance_t csip_instance;
+
+static uint8_t csip_notify_handler(struct bt_conn *conn,
+				  struct bt_gatt_subscribe_params *params,
+				  const void *data, uint16_t length)
+{
+	printk("%p\n", params);
+
+	if (params == &csip_instance.csip_set_lock_sub) {
+		printk("Received sink_csip_sub notification\n");
+		SET_FLAG(flag_all_notifications_received);
+	}
+
+	return BT_GATT_ITER_CONTINUE;
+}
+
+static uint8_t discover_csip_set_lock(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+			     struct bt_gatt_discover_params *params)
+{
+	struct bt_gatt_subscribe_params *subscribe_params;
+	int err;
+
+	if (!attr) {
+		printk("Discover complete\n");
+		(void)memset(params, 0, sizeof(*params));
+		return BT_GATT_ITER_STOP;
+	}
+
+	printk("					CSIP\n");
+
+	if (!bt_uuid_cmp(params->uuid, BT_UUID_CSIS_SET_LOCK)) {
+		printk("CSIP Set Lock Characteristic handle at %d\n", attr->handle);
+		subscribe_params = &csip_instance.csip_set_lock_sub;
+		memcpy(&uuid, BT_UUID_GATT_CCC, sizeof(uuid));
+		csip_instance.discover_params.uuid = &uuid.uuid;
+		csip_instance.discover_params.start_handle = attr->handle + 2;
+		csip_instance.discover_params.type = BT_GATT_DISCOVER_DESCRIPTOR;
+		subscribe_params->value_handle = bt_gatt_attr_value_handle(attr);
+
+		err = bt_gatt_discover(conn, &csip_instance.discover_params);
+		if (err) {
+			printk("Discover failed (err %d)\n", err);
+		}
+	} else if (!bt_uuid_cmp(params->uuid, BT_UUID_GATT_CCC)) {
+		printk("CSIP CCC handle at %d\n", attr->handle);
+		subscribe_params = &csip_instance.csip_set_lock_sub;
+		subscribe_params->notify = csip_notify_handler;
+		subscribe_params->value = BT_GATT_CCC_NOTIFY;
+		subscribe_params->ccc_handle = attr->handle;
+
+		err = bt_gatt_subscribe(conn, subscribe_params);
+		if (err && err != -EALREADY) {
+			printk("Subscribe failed (err %d)\n", err);
+		} else {
+			SET_FLAG(flag_csip_set_lock_discovered);
+			printk("[SUBSCRIBED]\n");
+		}
+	} else {
+		printk("Unknown handle at %d\n", attr->handle);
+		return BT_GATT_ITER_CONTINUE;
+	}
+
+	return BT_GATT_ITER_STOP;
+}
+
+static void discover_and_subscribe_csip_lock(void)
+{
+	int err = 0;
+
+	printk("%s\n", __func__);
+	printk("									CSIP\n");
+
+	memcpy(&uuid, BT_UUID_CSIS_SET_LOCK, sizeof(uuid));
+	csip_instance.discover_params.uuid = &uuid.uuid;
+	csip_instance.discover_params.start_handle = BT_ATT_FIRST_ATTRIBUTE_HANDLE;
+	csip_instance.discover_params.end_handle = BT_ATT_LAST_ATTRIBUTE_HANDLE;
+	csip_instance.discover_params.type = BT_GATT_DISCOVER_CHARACTERISTIC;
+	csip_instance.discover_params.func = discover_csip_set_lock;
+
+	err = bt_gatt_discover(default_conn, &csip_instance.discover_params);
+	if (err != 0) {
+		FAIL("Service Discovery failed (err %d)\n", err);
+		return;
+	}
+}
```

Comment:
I assume that you refer to using f.x. `bt_csip_set_coordinator_discover` in the `csip_set_coordinator.c` ?
If I would do that, then I would not have access setting `SET_FLAG(flag_csip_set_lock_discovered);`.

Reply:
In the discover callback you get the `struct bt_csip_set_coordinator_set_member *member`, which contains one or more `struct bt_csip_set_coordinator_csis_inst` which has the `struct bt_csip_set_coordinator_set_info` which then has the `bool lockable`, which you can use to determine whether or not the set lock characteristic has been found. 

If there is something missing from the implementation, it should be fixed in the implementation, rather than creating a new implementation :)

Reply:
Thats fair enough. I wasn't thinking of this being something missing from implementation, since the test actually doesn't test the client functionality, only that the server behaves as it should. There is ofc. more value in adding our client as well here if possible.

Ill make adjustment according to your comments.